From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Otto=20Pfl=C3=BCger?= <otto.pflueger@abscue.de>
Date: Mon, 9 Dec 2024 06:42:37 +0100
Subject: [PATCH 097/225] drm/panel: Add driver for ICNL9916 720x1600 panel
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Add a driver for the 720x1600 panel with an ICNL9916 controller found
on the Reeder S19 Max Pro S smartphone.

Signed-off-by: Otto Pflüger <otto.pflueger@abscue.de>
---
 drivers/gpu/drm/panel/Kconfig                 |  10 +
 drivers/gpu/drm/panel/Makefile                |   1 +
 .../panel/panel-chipone-icnl9916-720x1600.c   | 304 ++++++++++++++++++
 3 files changed, 315 insertions(+)
 create mode 100644 drivers/gpu/drm/panel/panel-chipone-icnl9916-720x1600.c

diff --git a/drivers/gpu/drm/panel/Kconfig b/drivers/gpu/drm/panel/Kconfig
index d592f4f4b..c991c5d77 100644
--- a/drivers/gpu/drm/panel/Kconfig
+++ b/drivers/gpu/drm/panel/Kconfig
@@ -105,6 +105,16 @@ config DRM_PANEL_BOE_TV101WUM_LL2
 	  Say Y here if you want to support for BOE TV101WUM-LL2
 	  WUXGA PANEL DSI Video Mode panel
 
+config DRM_PANEL_CHIPONE_ICNL9916_720_1600
+	tristate "Chipone ICNL9916 720x1600 panel"
+	depends on OF
+	depends on DRM_MIPI_DSI
+	depends on BACKLIGHT_CLASS_DEVICE
+	help
+	  Say Y here if you want to enable support for the Chipone ICNL9916
+	  MIPI-DSI video mode panel with a resolution of 720x1600.
+	  This type of panel is mainly found on mobile phones.
+
 config DRM_PANEL_EBBG_FT8719
 	tristate "EBBG FT8719 panel driver"
 	depends on OF
diff --git a/drivers/gpu/drm/panel/Makefile b/drivers/gpu/drm/panel/Makefile
index a4291dc39..1d869982b 100644
--- a/drivers/gpu/drm/panel/Makefile
+++ b/drivers/gpu/drm/panel/Makefile
@@ -9,6 +9,7 @@ obj-$(CONFIG_DRM_PANEL_BOE_TD4320) += panel-boe-td4320.o
 obj-$(CONFIG_DRM_PANEL_BOE_TH101MB31UIG002_28A) += panel-boe-th101mb31ig002-28a.o
 obj-$(CONFIG_DRM_PANEL_BOE_TV101WUM_LL2) += panel-boe-tv101wum-ll2.o
 obj-$(CONFIG_DRM_PANEL_BOE_TV101WUM_NL6) += panel-boe-tv101wum-nl6.o
+obj-$(CONFIG_DRM_PANEL_CHIPONE_ICNL9916_720_1600) += panel-chipone-icnl9916-720x1600.o
 obj-$(CONFIG_DRM_PANEL_DSI_CM) += panel-dsi-cm.o
 obj-$(CONFIG_DRM_PANEL_LVDS) += panel-lvds.o
 obj-$(CONFIG_DRM_PANEL_SIMPLE) += panel-simple.o
diff --git a/drivers/gpu/drm/panel/panel-chipone-icnl9916-720x1600.c b/drivers/gpu/drm/panel/panel-chipone-icnl9916-720x1600.c
new file mode 100644
index 000000000..040a98414
--- /dev/null
+++ b/drivers/gpu/drm/panel/panel-chipone-icnl9916-720x1600.c
@@ -0,0 +1,304 @@
+// SPDX-License-Identifier: GPL-2.0-only
+// Copyright (c) 2024 Otto Pflüger
+// Based on command information from the vendor device tree
+
+#include <linux/delay.h>
+#include <linux/mod_devicetable.h>
+#include <linux/module.h>
+#include <linux/reset.h>
+
+#include <drm/drm_mipi_dsi.h>
+#include <drm/drm_modes.h>
+#include <drm/drm_panel.h>
+#include <drm/drm_probe_helper.h>
+
+struct icnl9916_panel {
+	struct drm_panel panel;
+	struct mipi_dsi_device *dsi;
+	struct reset_control *reset;
+};
+
+static inline
+struct icnl9916_panel *to_icnl9916_panel(struct drm_panel *panel)
+{
+	return container_of(panel, struct icnl9916_panel, panel);
+}
+
+static void icnl9916_panel_reset(struct icnl9916_panel *ctx)
+{
+	reset_control_deassert(ctx->reset);
+	usleep_range(10000, 11000);
+	reset_control_assert(ctx->reset);
+	usleep_range(10000, 11000);
+	reset_control_deassert(ctx->reset);
+	msleep(120);
+}
+
+static int icnl9916_panel_on(struct icnl9916_panel *ctx)
+{
+	struct mipi_dsi_multi_context dsi_ctx = { .dsi = ctx->dsi };
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xf0, 0x99, 0x16, 0x0c);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xc1,
+				     0x00, 0x20, 0x20, 0xb4, 0x04, 0x30, 0x30,
+				     0x04, 0x40, 0x06, 0x22, 0x70, 0x33, 0x31,
+				     0x07, 0x11, 0x84, 0x4c, 0x00, 0x93, 0x13,
+				     0x63, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				     0x00, 0x00, 0x00, 0x00);
+	mipi_dsi_generic_write_seq_multi(&dsi_ctx, 0xc2, 0x00);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xc3,
+				     0x06, 0x00, 0xff, 0x00, 0xff, 0x4d, 0x10,
+				     0x01);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xc4,
+				     0x04, 0x33, 0xb8, 0x40, 0x00, 0xbc, 0x00,
+				     0x00, 0x00, 0x00, 0x00, 0xf0);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xc5,
+				     0x03, 0x21, 0x96, 0xc8, 0x3e, 0x00, 0x04,
+				     0x01, 0x14, 0x04, 0x0e, 0x18, 0xc6, 0x03,
+				     0x64, 0xff, 0x01, 0x04, 0x18, 0x22, 0x45,
+				     0x14, 0x38);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xc6,
+				     0x72, 0x24, 0x13, 0x2b, 0x2b, 0x28, 0x3f,
+				     0x02, 0x16, 0x16, 0x00, 0x01);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xca,
+				     0x34, 0x50, 0x04, 0x19, 0x46, 0x94, 0x41,
+				     0x8f, 0x44, 0x44, 0x36, 0x50, 0x54, 0x54,
+				     0x39, 0x5a, 0x5a, 0x5a, 0x33, 0x00, 0x01,
+				     0x01, 0x0e, 0x3f, 0xd2, 0x00, 0x05, 0x00,
+				     0x00, 0x5a, 0x5a, 0x00);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xb2,
+				     0x05, 0x04, 0x10, 0x10, 0x44, 0x44, 0x82,
+				     0x88, 0x44, 0x86, 0x84, 0x86, 0x84, 0x86,
+				     0x84, 0x86, 0x84, 0x86, 0x84, 0x86, 0x84,
+				     0x00, 0x00, 0x00);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xb3,
+				     0xf4, 0x01, 0x01, 0x11, 0x91, 0x86, 0x00,
+				     0x00, 0x84, 0x00, 0x00, 0x65, 0x4a, 0x65,
+				     0x4a);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xb4,
+				     0x19, 0x0b, 0x06, 0x0b, 0x06, 0x26, 0x26,
+				     0x88, 0xa2, 0x88, 0x44, 0x3b, 0x26, 0x00,
+				     0x55, 0x3c, 0x02, 0x08, 0x20, 0x30, 0x00,
+				     0x12, 0x20, 0x40, 0x11, 0x10, 0x20, 0x00);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xb5,
+				     0x00, 0x00, 0x08, 0x04, 0x2e, 0x2f, 0x0c,
+				     0x0e, 0x10, 0x12, 0x14, 0x16, 0x18, 0x1a,
+				     0x28, 0x02, 0x02, 0x02, 0x02, 0x02, 0x02,
+				     0x02, 0xff, 0xff, 0xfc, 0x0c, 0x00, 0x00,
+				     0x3c, 0x00);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xb6,
+				     0x00, 0x00, 0x09, 0x05, 0x2e, 0x2f, 0x0d,
+				     0x0f, 0x11, 0x13, 0x15, 0x17, 0x19, 0x1b,
+				     0x29, 0x02, 0x02, 0x02, 0x02, 0x02, 0x02,
+				     0x02, 0xff, 0xff, 0xfc, 0x0c, 0x00, 0x00,
+				     0x3c, 0x00);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xc0,
+				     0x40, 0x93, 0xff, 0xff, 0xff, 0x3f, 0xff,
+				     0x00, 0xff, 0x00, 0xcc, 0xb1, 0x23, 0x45,
+				     0x67, 0x89, 0xad, 0xff, 0xff, 0xf0);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xb7,
+				     0x0a, 0x0a, 0x0a, 0x0a, 0x0a, 0x0a, 0x0a,
+				     0x0a, 0x0a, 0x0a, 0x0a, 0x0a, 0x0a, 0x0a,
+				     0x0a, 0x0a, 0x0a, 0x0a, 0x0a, 0x0a, 0x0a,
+				     0x0a);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xbc,
+				     0x00, 0x00, 0x00, 0x00, 0x04, 0x00, 0xff,
+				     0xf0, 0x0b, 0x33, 0x5c, 0x5b, 0x43, 0x33,
+				     0x00, 0x5a, 0x5a, 0x55, 0x55);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xbd, 0xa1, 0x0a, 0x52, 0xa6);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xbf,
+				     0x0c, 0x19, 0x0c, 0x19, 0x00, 0x11, 0x04,
+				     0x18, 0x50);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xc7,
+				     0x77, 0x77, 0x77, 0x77, 0x77, 0x77, 0x77,
+				     0x77, 0x70, 0x77, 0x77, 0x77, 0x77, 0x77,
+				     0x77, 0x77, 0x77, 0x70, 0x31, 0x00, 0x01,
+				     0xff, 0xff, 0x00, 0x8a, 0x8a, 0x40);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x80,
+				     0xff, 0xfa, 0xf3, 0xec, 0xe6, 0xe0, 0xdb,
+				     0xd7, 0xd3, 0xc5, 0xba, 0xb0, 0xa8, 0xa1,
+				     0x9a, 0x8f, 0x85, 0x7b, 0x73, 0x72, 0x6a,
+				     0x61, 0x58, 0x4e, 0x42, 0x32, 0x28, 0x1b,
+				     0x17, 0x13, 0x0f, 0x0a);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x81,
+				     0xff, 0xfa, 0xf3, 0xec, 0xe6, 0xe0, 0xdb,
+				     0xd7, 0xd3, 0xc5, 0xba, 0xb0, 0xa8, 0xa1,
+				     0x9a, 0x8f, 0x85, 0x7b, 0x73, 0x72, 0x6a,
+				     0x61, 0x58, 0x4e, 0x42, 0x32, 0x28, 0x1b,
+				     0x17, 0x13, 0x0f, 0x0a);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x82,
+				     0xff, 0xfa, 0xf3, 0xec, 0xe6, 0xe0, 0xdb,
+				     0xd7, 0xd3, 0xc5, 0xba, 0xb0, 0xa8, 0xa1,
+				     0x9a, 0x8f, 0x85, 0x7b, 0x73, 0x72, 0x6a,
+				     0x61, 0x58, 0x4e, 0x42, 0x32, 0x28, 0x1b,
+				     0x17, 0x13, 0x0f, 0x0a);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x83,
+				     0x01, 0x06, 0x02, 0x00, 0x00, 0x06, 0x02,
+				     0x00, 0x00, 0x06, 0x02, 0x00, 0x00, 0x0a,
+				     0x06, 0x02, 0x00, 0x0a, 0x06, 0x02, 0x00,
+				     0x0a, 0x06, 0x02, 0x00);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x84,
+				     0x30, 0x3d, 0x13, 0x98, 0x32, 0x69, 0x01,
+				     0xa7, 0x64, 0x30, 0x3d, 0x13, 0x98, 0x32,
+				     0x69, 0x01, 0xa7, 0x64, 0x30, 0x3d, 0x13,
+				     0x98, 0x32, 0x69, 0x01, 0xa7, 0x64);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x89, 0xfe, 0xfe, 0x00);
+	mipi_dsi_generic_write_seq_multi(&dsi_ctx, 0x35, 0x00);
+	mipi_dsi_generic_write_seq_multi(&dsi_ctx, 0x11, 0x00);
+	mipi_dsi_msleep(&dsi_ctx, 120);
+	mipi_dsi_generic_write_seq_multi(&dsi_ctx, 0x29, 0x00);
+	mipi_dsi_usleep_range(&dsi_ctx, 10000, 11000);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xbd, 0xa1, 0x0a, 0x52, 0xae);
+	mipi_dsi_generic_write_seq_multi(&dsi_ctx, 0x6d, 0x02);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xf0, 0x00, 0x00, 0x00);
+
+	return dsi_ctx.accum_err;
+}
+
+static int icnl9916_panel_off(struct icnl9916_panel *ctx)
+{
+	struct mipi_dsi_multi_context dsi_ctx = { .dsi = ctx->dsi };
+
+	mipi_dsi_generic_write_seq_multi(&dsi_ctx, 0x28);
+	mipi_dsi_msleep(&dsi_ctx, 50);
+	mipi_dsi_generic_write_seq_multi(&dsi_ctx, 0x10);
+	mipi_dsi_msleep(&dsi_ctx, 120);
+
+	return dsi_ctx.accum_err;
+}
+
+static int icnl9916_panel_prepare(struct drm_panel *panel)
+{
+	struct icnl9916_panel *ctx = to_icnl9916_panel(panel);
+	struct device *dev = &ctx->dsi->dev;
+	int ret;
+
+	icnl9916_panel_reset(ctx);
+
+	ret = icnl9916_panel_on(ctx);
+	if (ret < 0) {
+		dev_err(dev, "Failed to initialize panel: %d\n", ret);
+		reset_control_assert(ctx->reset);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int icnl9916_panel_unprepare(struct drm_panel *panel)
+{
+	struct icnl9916_panel *ctx = to_icnl9916_panel(panel);
+	struct device *dev = &ctx->dsi->dev;
+	int ret;
+
+	ret = icnl9916_panel_off(ctx);
+	if (ret < 0)
+		dev_err(dev, "Failed to un-initialize panel: %d\n", ret);
+
+	reset_control_assert(ctx->reset);
+
+	return 0;
+}
+
+static const struct drm_display_mode icnl9916_panel_mode = {
+	.clock = 96000,
+	.hdisplay = 720,
+	.hsync_start = 720 + 70,
+	.hsync_end = 720 + 70 + 4,
+	.htotal = 720 + 70 + 4 + 80,
+	.vdisplay = 1600,
+	.vsync_start = 1600 + 180,
+	.vsync_end = 1600 + 180 + 4,
+	.vtotal = 1600 + 180 + 4 + 32,
+	.width_mm = 68,
+	.height_mm = 151,
+	.type = DRM_MODE_TYPE_DRIVER,
+};
+
+static int icnl9916_panel_get_modes(struct drm_panel *panel,
+				       struct drm_connector *connector)
+{
+	return drm_connector_helper_get_modes_fixed(connector, &icnl9916_panel_mode);
+}
+
+static const struct drm_panel_funcs icnl9916_panel_panel_funcs = {
+	.prepare = icnl9916_panel_prepare,
+	.unprepare = icnl9916_panel_unprepare,
+	.get_modes = icnl9916_panel_get_modes,
+};
+
+static int icnl9916_panel_probe(struct mipi_dsi_device *dsi)
+{
+	struct device *dev = &dsi->dev;
+	struct icnl9916_panel *ctx;
+	int ret;
+
+	ctx = devm_kzalloc(dev, sizeof(*ctx), GFP_KERNEL);
+	if (!ctx)
+		return -ENOMEM;
+
+	ctx->reset = devm_reset_control_get_shared(dev, NULL);
+	if (IS_ERR(ctx->reset))
+		return dev_err_probe(dev, PTR_ERR(ctx->reset),
+				     "Failed to get chip reset\n");
+
+	ctx->dsi = dsi;
+	mipi_dsi_set_drvdata(dsi, ctx);
+
+	dsi->lanes = 4;
+	dsi->format = MIPI_DSI_FMT_RGB888;
+	dsi->mode_flags = MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST |
+			  MIPI_DSI_MODE_VIDEO_HSE;
+	dsi->hs_rate = 691000000;
+	dsi->lp_rate = 20000000;
+
+	drm_panel_init(&ctx->panel, dev, &icnl9916_panel_panel_funcs,
+		       DRM_MODE_CONNECTOR_DSI);
+	ctx->panel.prepare_prev_first = true;
+
+	ret = drm_panel_of_backlight(&ctx->panel);
+	if (ret)
+		return dev_err_probe(dev, ret, "Failed to get backlight\n");
+
+	drm_panel_add(&ctx->panel);
+
+	ret = mipi_dsi_attach(dsi);
+	if (ret < 0) {
+		drm_panel_remove(&ctx->panel);
+		return dev_err_probe(dev, ret, "Failed to attach to DSI host\n");
+	}
+
+	return 0;
+}
+
+static void icnl9916_panel_remove(struct mipi_dsi_device *dsi)
+{
+	struct icnl9916_panel *ctx = mipi_dsi_get_drvdata(dsi);
+	int ret;
+
+	ret = mipi_dsi_detach(dsi);
+	if (ret < 0)
+		dev_err(&dsi->dev, "Failed to detach from DSI host: %d\n", ret);
+
+	drm_panel_remove(&ctx->panel);
+}
+
+static const struct of_device_id icnl9916_panel_of_match[] = {
+	{ .compatible = "chipone,icnl9916-720x1600-panel" },
+	{ /* sentinel */ }
+};
+MODULE_DEVICE_TABLE(of, icnl9916_panel_of_match);
+
+static struct mipi_dsi_driver icnl9916_panel_driver = {
+	.probe = icnl9916_panel_probe,
+	.remove = icnl9916_panel_remove,
+	.driver = {
+		.name = "panel-icnl9916-720x1600",
+		.of_match_table = icnl9916_panel_of_match,
+	},
+};
+module_mipi_dsi_driver(icnl9916_panel_driver);
+
+MODULE_AUTHOR("Otto Pflüger <otto.pflueger@abscue.de>");
+MODULE_DESCRIPTION("DRM driver for ICNL9916 720x1600 video mode DSI panel");
+MODULE_LICENSE("GPL");
